### PR TITLE
[CARBONDATA-2738]Block Preaggregate, Compaction, Dictionary Exclude/Include for child columns for Complex datatype

### PR DESCRIPTION
**Pull request by Indhumathi27**
Block Preaggregate, Compaction, Dictionary Exclude/Include for child columns and Update Complex datatype columns  for Complex datatype

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done
       Test cases added
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7071/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5846/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7076/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5853/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5813/

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 1]
```diff
@@ -720,18 +720,18 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
         tableProperties.get(CarbonCommonConstants.DICTIONARY_EXCLUDE).get.split(',').map(_.trim)
       dictExcludeCols
         .foreach { dictExcludeCol =>
+          if (dictExcludeCol.contains(".") || dictExcludeCol.contains("[")) {
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: do not check, for contains ".", get from field whether it is child column, then you can check whether that is given in include or exclude and throw exception

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 2]
```diff
@@ -748,6 +748,10 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
       dictIncludeCols =
         tableProperties(CarbonCommonConstants.DICTIONARY_INCLUDE).split(",").map(_.trim)
       dictIncludeCols.foreach { distIncludeCol =>
+        if (distIncludeCol.contains(".") || distIncludeCol.contains("[")) {
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: same as above

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala [hunk 3]
```diff
@@ -59,6 +60,13 @@ private[sql] case class CarbonProjectForUpdateCommand(
       return Seq.empty
     }
     val carbonTable = CarbonEnv.getCarbonTable(databaseNameOp, tableName)(sparkSession)
+    columns.foreach { col =>
+      var dataType = carbonTable.getColumnByName(tableName, col).getColumnSchema.getDataType
+      if(dataType.getName.equals("STRUCT") || dataType.getName.equals("ARRAY")) {
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala]: do not check for hard coded value, get the datatype and dicrectly check `isComplexType` and throw error

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5824/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5828/

**Comment by Indhumathi27** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7172/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5948/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5846/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5847/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7185/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5959/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7199/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5974/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5865/

**Comment by ajantha-bhat** [comment]
check and handle check style for all the new changes, example: CarbonProjectForUpdateCommand.scala

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala [hunk 4]
```diff
@@ -22,7 +22,7 @@ import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project}
 import org.apache.spark.sql.execution.command._
 import org.apache.spark.sql.execution.command.management.CarbonLoadDataCommand
 import org.apache.spark.sql.execution.datasources.LogicalRelation
-import org.apache.spark.sql.types.ArrayType
+import org.apache.spark.sql.types.{ArrayType, StructType}
```
Comment (ajantha-bhat) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala]: might be redundant import. run check style manually.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5995/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5999/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5881/

**Comment by ajantha-bhat** [comment]
LGTM

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAlterTableCompactionCommand.scala [hunk 5]
```diff
@@ -82,6 +82,12 @@ case class CarbonAlterTableCompactionCommand(
       throw new MalformedCarbonCommandException("Unsupported operation on non transactional table")
     }
 
+    if (table.getTableInfo.getFactTable.getListOfColumns.asScala
+      .exists(m => m.getDataType.isComplexType)) {
+      throw new MalformedCarbonCommandException(
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonAlterTableCompactionCommand.scala]: throw UnsupportedOperationException

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala [hunk 6]
```diff
@@ -59,6 +60,13 @@ private[sql] case class CarbonProjectForUpdateCommand(
       return Seq.empty
     }
     val carbonTable = CarbonEnv.getCarbonTable(databaseNameOp, tableName)(sparkSession)
+    columns.foreach { col =>
+      val dataType = carbonTable.getColumnByName(tableName, col).getColumnSchema.getDataType
+      if (dataType.isComplexType) {
+        throw new MalformedCarbonCommandException("Unsupported operation on Complex data type")
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForUpdateCommand.scala]: throw UnsupportedOperationException

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateUtil.scala [hunk 7]
```diff
@@ -354,7 +354,13 @@ object PreAggregateUtil {
         !expression.isInstanceOf[AttributeReference]) {
       newColumnName
     } else {
-      expression.asInstanceOf[AttributeReference].name
+      if (expression.isInstanceOf[GetStructField] || expression.isInstanceOf[GetArrayItem]) {
+        throw new MalformedCarbonCommandException(
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/preaaggregate/PreAggregateUtil.scala]: throw UnsupportedOperationException

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala [hunk 8]
```diff
@@ -712,5 +713,166 @@ class TestComplexDataType extends QueryTest with BeforeAndAfterAll {
...
+    sql(
+      "create table table1 (roll int,a struct<b:int,c:string,d:int,e:string,f:struct<g:int," +
+      "h:string,i:int>,j:int>) stored " +
+      "by " +
+      "'carbondata' tblproperties('dictionary_exclude'='a.b')"))
+    assertResult(
+      "DICTIONARY_EXCLUDE column: a.b does not exist in table or unsupported for complex child " +
+      "column. Please check create table statement.")(
+      structException.getMessage)
+    sql("DROP TABLE IF EXISTS table1")
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala]: drop will not be executred if assertion fails. Please make sure all tables are deleted in afterAll also.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala [hunk 9]
```diff
@@ -712,5 +713,166 @@ class TestComplexDataType extends QueryTest with BeforeAndAfterAll {
...
+      "/Struct.csv' into table table1 options('delimiter'=','," +
+      "'quotechar'='\"','fileheader'='roll,person','complex_delimiter_level_1'='$'," +
+      "'complex_delimiter_level_2'='&')")
+    val exception = intercept[MalformedCarbonCommandException](
+      sql("alter table table1 compact 'major'").show(false))
+    assertResult(
+      "Compaction is unsupported for Table containing Complex Columns")(
+      exception.getMessage)
+    val exception1 = intercept[MalformedCarbonCommandException](
+      sql("alter table table1 compact 'minor'").show(false))
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/complexType/TestComplexDataType.scala]: remove .show from from all queries

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7242/

**Comment by Indhumathi27** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6016/

**Comment by brijoobopanna** [comment]
retest sdv please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7253/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5889/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7267/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6036/

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**